components/accounts/src/lib.rs

```diff
@@ -0,0 +1,349 @@
+        self.state
+    }
+
+    pub fn export_persisted_state(&self) -> Result<String> {
+        self.account.to_json()
+    }
+
+    pub fn update_devices(&mut self) -> DeviceConstellation {
+        match self.account.get_devices() {
+            Ok(devices) => {
+                self.devices = devices;
+            }
+            Err(e) => match e.kind() {
+                ErrorKind::RemoteError { code: 401, .. } => {
+                    self.state = State::AuthenticationProblem;
+                    self.devices = vec![];
+                }
+                _ => {}
+            },
+        };
+        self.get_devices()
+    }
+
+    pub fn get_devices(&self) -> DeviceConstellation {
+        let (mut own_devices, other_devices): (Vec<_>, Vec<_>) = self
+            .devices
+            .iter()
+            .map(|d| d.clone())
+            .partition(|d| d.is_current_device);
+        let current_device = match own_devices.len() {
+            0 => None,
+            1 => Some(own_devices.remove(0)),
+            _ => {
+                log::error!("Found multiple own devices?!");
+                Some(own_devices.remove(0))
+            }
+        };
+        DeviceConstellation {
+            current_device,
+            other_devices,
+        }
+    }
+
+    // Escape hatch used by the FFI crate. Do not use!
+    pub fn get_account(&mut self) -> &mut FirefoxAccount {
+        &mut self.account
+    }
+
+    /// State transition matrix.
+    /// If state+event results in a transition, a state
+    /// is returned.
+    fn next_state(state: &State, event: &Event) -> Option<State> {
+        match state {
+            State::Start => match event {
+                Event::Init { .. } => Some(State::Start),
+                Event::AccountRestored => Some(State::Authenticated),
+                Event::AccountCreated => Some(State::NotAuthenticated),
+                _ => None,
+            },
+            State::NotAuthenticated => match event {
+                Event::FailedToAuthenticate => Some(State::NotAuthenticated),
+                Event::CompletedAuthFlow { .. } => Some(State::Authenticated),
+                _ => None,
+            },
+            State::Authenticated => match event {
+                Event::AuthenticationError => Some(State::AuthenticationProblem),
+                Event::Logout => Some(State::NotAuthenticated),
+                _ => None,
+            },
+            State::AuthenticationProblem => match event {
+                Event::FailedToAuthenticate => Some(State::AuthenticationProblem),
+                Event::RecoveredFromAuthenticationProblem => Some(State::Authenticated),
+                Event::CompletedAuthFlow { .. } => Some(State::Authenticated),
+                Event::Logout => Some(State::NotAuthenticated),
+                _ => None,
+            },
+        }
+    }
+
+    fn state_actions(&mut self, via: &Event) -> Option<Event> {
+        match self.state {
```

Comment:
FWIW, I personally find this "big nested switch statements" approach to state machines a bit unwieldy, especially as the set of states and events grows over time. Did you consider an approach where each state is a separate "class" (or I guess enum variant?) with methods to call for the various events.  I'm thinking along the lines of what we did for [fxa-pairing-channel](https://github.com/mozilla/fxa-pairing-channel/blob/master/src/states.js) but I totally buy the suggestion that it's an awkward fit for rust.